### DISCUSSION TITLE
Replaying up to a certain point in time

### DISCUSSION THREAD

**Discussion by Ahrengot** [spatie/laravel-event-sourcing#417] [Q&A]
One recurring advantage often attributed to event sourcing is the ability to replay events up to a specific moment in time. For instance, you could inquire about the state of your account at precisely 9:47 AM on December 12th, 2020.

This concept is mentioned by Greg Young in his presentation here https://www.youtube.com/watch?v=8JKjvY4etTY. It's also described in the No Plans to Merge episode where Daniel pitches event sourcing: https://noplanstomerge.com/115

Having gone through the documentation, and the source code, I haven't come across any explicit reference to replaying events up to a given point in time. While the artisan replay command does offer a `--from` parameter to specify the starting event ID, it lacks a corresponding `--to` parameter.

Have overlooked this feature? Has there been any prior discussions around this? I found issue #77 to be somewhat related, but it does not delve into implementation specifics.

If not, I'm curious if there is any community support or interest in adding this feature. I wouldn't mind taking a shot at a PR myself, but want to make sure this hasn't already been turned down in the past and if others actually need this.

**Comment by sebastiandedeyne** [comment]
Hi @Ahrengot,

Thanks for the question, it inspired me to write a post on the topic! 😄 https://sebastiandedeyne.com/event-sourcing/replaying-events/

As for our library, the closest thing we have are event queries which allow you to build state for a specific period in time: https://spatie.be/docs/laravel-event-sourcing/v7/advanced-usage/event-queries

The replay command is not meant to be used to build projections at a certain point in time, because that would cause your projections to be in a different state than your aggregate roots. This can be fine if it's explicit, but depends on a lot of factors (as written in my blog post).

If you have suggestions for this package to make it easier to deal with past data, we're definitely open for them!

**Reply by Ahrengot** [reply]
Thanks for the blog post and I also caught your interview on PHP Annotated (https://www.youtube.com/watch?v=qTUB_jESAKY) which gave more clarity.

I should have been more clear above, in that I'm not looking to time travel in production, but in development. Either when working on a new feature or if production somehow got into an unexpected state, it would be nice to pull the database and time travel locally  to see how the app changed over time. Sort of like a git bisect.

I would also assume both aggregate roots and projectors would be able to time travel, but maybe I'm missing some context here?
